pir/cpp/client_test.cpp

```diff
@@ -72,12 +77,50 @@ TEST_F(PIRClientTest, TestProcessResponse) {
   Context()->Encoder()->encode(value, pt);
   vector<Ciphertext> ct(1);
   Encryptor()->encrypt(pt, ct[0]);
-  PIRPayload payload = PIRPayload::Load(ct);
 
-  auto result = client_->ProcessResponse(payload).ValueOrDie();
+  Response reply;
+  SaveCiphertexts(ct, reply.mutable_reply());
+
+  auto result = client_->ProcessResponse(reply).ValueOrDie();
   ASSERT_EQ(result, value);
 }
 
+TEST_F(PIRClientTest, TestRequestSerialization) {
+  int64_t value = 987654321;
+  Plaintext pt;
+  Context()->Encoder()->encode(value, pt);
+  vector<Ciphertext> ct(1);
+  Encryptor()->encrypt(pt, ct[0]);
+
+  Response reply_proto;
+  SaveCiphertexts(ct, reply_proto.mutable_reply());
+
+  auto reloaded = LoadCiphertexts(Context()->SEALContext(), reply_proto.reply())
+                      .ValueOrDie();
+
+  ASSERT_EQ(reloaded.size(), 1);
+
+  auto keygen_ = make_unique<KeyGenerator>(Context()->SEALContext());
+  auto elts = generate_galois_elts(DEFAULT_POLY_MODULUS_DEGREE);
+  GaloisKeys gal_keys = keygen_->galois_keys_local(elts);
+
+  Request proto_query;
+  SaveCiphertexts(ct, proto_query.mutable_query());
+  auto encoded_keys = SEALSerialize<GaloisKeys>(gal_keys).ValueOrDie();
+  proto_query.set_keys(encoded_keys);
+
+  auto request = LoadCiphertexts(Context()->SEALContext(), proto_query.query())
+                     .ValueOrDie();
+  auto keys =
+      SEALDeserialize<GaloisKeys>(Context()->SEALContext(), proto_query.keys())
+          .ValueOrDie();
+
+  ASSERT_EQ(request.size(), 1);
+  for (auto& elt : elts) {
+    ASSERT_TRUE(keys.has_key(elt));
```

Comment:
There are much better asserts in gmock. I'd suggest using ElementsAreArray here.

Reply:
Not sure how should I use ElementsAreArray in this particular scenario.

Reply:
Could you not do something like:
`ASSERT_THAT(keys, ElementsAreArray(elts));`